joomla/joomla-cms
# [RFC] An approach to soft upgrade bootstrap

The shift to a new Bootstrap major version is very difficult for Joomla and can be done only on a new major version like Joomla 4, becaus it breaks backwards compatibility. This would then require for all Joomla 3.x sites a new template when they want to upgrade to Joomla 4. Extension developers need to create special overrides, two different versions of their software or are forced to load their own namespaced version of frontend framework which has an impact on site speed.

This PR proposes a 100% backwards compatible way for a switch to a new frontend framework (FF) within the same major version of Joomla. All framework related code will be pushed to their own layout files where it belongs to. This means that Joomla can hold two different versions of frontend framework at the same time (only **ONE** will be loaded during a page load). The deprecated one and the new one. In practice it would mean that in Joomla 3.x, BS2 will be deprecated and BS 3 or 4 introduced, it will allow the site admins to move to BS3 or BS4 within Joomla 3. It gives the template and extension developers an unified way to deal with the problem and to switch faster to new frontend frameworks in Joomla core when required. Also the site upgrade to a new frontend framework can be done step wise and not with a big bang. You can have your old template and the new one installed at the same time and then migrate page by page by assigning the menu item to the new template, without doing any layout override.

In the Tauristar template project https://github.com/Digital-Peak-Incubator/tpl_tauristar, we made a proof of concept with that method. A BS3 template was created with no overrides. When Protostar is loaded, then only the default.php files are rendered, when Tauristar is activated then the bs3 related files like default.bs3.php are rendered. Mostly it is enough that the framework specific file does set some variables for the default file. So we have no duplicated code in Joomla as shown in this PR.

BS3 is used only for demonstration purposes, treat is a placeholder for the next frontend framework major version used by Joomla.
### Summary of Changes

The template can define a new hidden parameter against which framework it is built to, like:

```
<field
    type="hidden"
    name="framework"
    default="bs3"
/>
```

By loading the template params, the view, module helper and layout manager do search in their paths first for a framework specific file. If none is found, then they load the default file. If none parameter is set at all in the template, only the default files are loaded.
The params approach acts here only for demonstration purposes, there are many other ways how a template can define which framework it is made for.
### Testing Instructions
- Add the following code to the file templates/protostar/templateDetails.xml after the fluidContainer field

```
<field
    type="hidden"
    name="framework"
    default="bs3"
/>
```
- Save the template
- Create a new contact
- Create a contacts category menu item
- On the front, open the contacts link

The classes _col-md_ should be added to the grid cells and not _span_.
### Pros
- 100% BC
- Framework related file can output a complete new content or set some variables for the default layout
- No duplicate code
- Template doesn't need to be shipped with all the layouts and template files of all components
- Faster upgrade or switch to new frontend frameworks during one major version of Joomla
- Reducing complexity for template developers as they don't need to create overrides for all extensions
- Saver sites as it will no be neccessary anmore to make template/layout overrides just because of a different frontend framework, fixes will not get forgotten which are applied to the core, but not the override
- It is optional for the extension dev to create framework specific layouts and views
- 3rd party extensions and the core will use the same way to deal with different frontend frameworks
### Cons
- Adds a new layer of complexity to the core
- Performance decrease as more file lookups need to be done

At the moment it is a big mess in Joomla because most of the templates do load their own frontend framework or load bs3 with tons of overrides for Joomla core. The extension developers do the same or have to do some other quirks to work flawless with any template. I know it adds some complexity but I'm convinced it is worth it as it solves more problems than it produces. Joomla will be able then to adapt faster to new technologies and to do real one click major upgrades. If this problem, which is addressed by this PR is not solved, then we are repeating the same problem over and over again and didn't learn from the past.

I'm aware that this is more of a practical approach and goes against some architecture decisions, but it allows us to extract all the frontend framework related code into files. On a later stage, it will be possible to remove them from the core completely and to make them installable or whatever. But the target must be to disconnect the core from any frontend framework and this is the first step to it.
